File: MaxValInRow.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
/*
algorithm:
traverse level by level and at each level keep track of max element
after that level, add max element to result list

time complexity: O(n), since traversing all nodes
space complexity: O(n/2)
*/
class MaxValInRow {
    public List<Integer> largestValues(TreeNode root) {
        if(root == null){
            return new ArrayList<>();
        }
        
        List<Integer>result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        
        queue.add(root);
        
        while(!queue.isEmpty()){
            int size = queue.size();
             int max = Integer.MIN_VALUE;
            for(int i = 0 ;i < size; i++){
               
                TreeNode node = queue.poll();
                if(node.val > max){
                    max = node.val;
                }
                if(node.left != null)queue.add(node.left);
                if(node.right != null) queue.add(node.right);
            }
            
            result.add(max);
        }
        return result;
    }
}

File: MaxVal_In_Row.cpp
// Time Complexity : O(N) where N represents the no.of nodes in the tree
// Space Complexity : O(H) where H is the height of the tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Nope
// Problem Link: https://leetcode.com/problems/find-largest-value-in-each-tree-row/
// Submission Link: https://leetcode.com/submissions/detail/543018254/

class Solution {
    vector<int> result;
    int maxDepth;
public:
    vector<int> largestValues(TreeNode* root) {
        maxDepth = -1;
        helper(root, 0);
        
        return result;
    }
    
    void helper(TreeNode* root, int depth){
        if(root == NULL)
            return;
        
        if(depth > maxDepth){
            maxDepth = depth;
            result.push_back(root -> val);
        }
        else{
            result[depth] = max(root -> val, result[depth]);
        }
        
        helper(root -> left, depth + 1);
        helper(root -> right, depth + 1);
    }
};